src/hevm/src/EVM/UnitTest.hs

```diff
@@ -191,23 +202,100 @@ checkFailures UnitTestOptions { .. } method bailed = do
         in pure (shouldFail == failed)
       _ -> error "internal error: unexpected failure code"
 
--- | Randomly generates the calldata arguments and runs the test
-fuzzTest :: UnitTestOptions -> Text -> [AbiType] -> VM -> Property
-fuzzTest opts sig types vm = forAllShow (genAbiValue (AbiTupleType $ Vector.fromList types)) (show . ByteStringS . encodeAbiValue)
-  $ \args -> ioProperty $
-    fst <$> runStateT (EVM.Stepper.interpret (oracle opts) (runUnitTest opts sig args)) vm
+-- | Either generates the calldata by mutating an example from the corpus, or synthesizes a random example
+genWithCorpus :: UnitTestOptions -> Corpus -> W256 -> Text -> [AbiType] -> Gen AbiValue
+genWithCorpus opts corpus codeHash sig tps = do
+  case Map.lookup (codeHash, sig) corpus of
+    Nothing -> genAbiValue (AbiTupleType $ Vector.fromList tps)
+    Just examples -> frequency
+      [ (mutations opts,         Test.QuickCheck.elements (Map.elems examples) >>= mutateAbiValue)
+      , (100 - (mutations opts), genAbiValue (AbiTupleType $ Vector.fromList tps))
+      ]
+
+-- | Randomly generates the calldata arguments and runs the test, updates the corpus with a new example if we explored a new path
+fuzzTest :: UnitTestOptions -> Text -> [AbiType] -> VM -> StateT Corpus IO FuzzResult
+fuzzTest opts sig types vm = do
+  let codeHash' = _codehash . fromJust $ currentContract vm
+  corpus <- get
+  args <- liftIO . generate $ genWithCorpus opts corpus codeHash' sig types
+  (res, (vm', trace')) <- liftIO $ runStateT (interpretWithTraceId opts (runUnitTest opts sig args)) (vm, [])
+  modify $ updateCorpus (codeHash', sig) trace' args
+  if res
+     then pure Pass
+     else pure $ Fail vm' (show . ByteStringS . encodeAbiValue $ args)
+  where
+    updateCorpus k1 k2 v c = case Map.lookup k1 c of
+      Nothing -> Map.insert k1 (Map.insert k2 v mempty) c
+      Just m' -> Map.insert k1 (Map.insert k2 v m') c
+
+type TraceIdState = (VM, [(W256, Int)])
+
+-- | This interpreter is similar to interpretWithCoverage, except instead of
```

Comment:
This comment is a lie, the hash accumulator was an old idea that actually ended up slowing things down. However I did end up going with a much faster data structure for the traces. Instead of a `MultiSet OpLocation` where `OpLocation` is a `Contract` and a bytecode index, we store insted a list of `(codehash, bytecode index)`. 

Using a codehash keeps the size of the corpus on disk small, and using a list means we can get `O(1)` insertion of new elements by just consing them onto the head of the list (instead of `O(log n)` for a MultiSet). 

With these optimizations in place the extra overhead introduced by the serilization / instrumentation is somewhere between 10 and 15 percent.

Reply:
Note that the use of codehash means we'll need some extra state if we want to show pretty coverage reports to the user (perhaps a mapping from codehash to `SolcContract` or smth in the `Corpus`?)

Reply:
we probably want to rewrite the existing coverage based interpreter to used the new data structures before merging, but I was still just trying to move fast and figure out what worked best so I used a seperate one to start with.

Reply:
oh also it's worth noting that using a `MultiSet` or a list instead of a `Set` for the traces means that traces that get further into a loop will be treated as a new trace, which seems like a nice thing.